src/parser.rs

```diff
@@ -301,25 +301,33 @@ impl<'a> Parser<'a, Vec<Expression<'a>>> for ParseExpressionList {
     fn parse(&self, state: ParseState<'a>) -> Result<(ParseState<'a>, Vec<Expression<'a>>), ParseAbort> {
         let mut state = state;
         let mut expressions = Vec::new();
+        let mut on_comma = false;
 
         loop {
             match ParseExpression.parse(state) {
                 Ok((next_state, expression)) => {
                     expressions.push(expression);
                     state = next_state;
+                    on_comma = false;
                 },
                 Err(_) => break,
             }
 
             match (EatToken { kind: TokenKind::Operator(",") }.parse(state)) {
                 Ok((next_state, _)) => {
                     state = next_state;
+                    on_comma = true;
                 },
                 Err(_) => break,
             }
         }
 
-        Ok((state, expressions))
+        if on_comma {
+            Err(ParseAbort::NoMatch)
+        }
+        else {
+            Ok((state, expressions))
+        }
```

Comment:
There's actually a really handy way to solve this without introducing another piece of state! I use it in the one-or-more variant of this same mini-parser with a couple tweaks.

I think it works. Basically:
1. Try to parse an expression. If it fails, return success with an empty list (or an error in the one-or-more case)
2. Loop:
    1. Parse a comma. If it fails, break the loop.
    1. Parse an expression. If it fails, return an error ("expected expression after comma found X")
3. Return success with all of the expressions